service/src/event/ordering_task.rs

```diff
@@ -0,0 +1,730 @@
+}
+
+#[derive(Debug)]
+pub struct OrderingTask {}
+
+impl OrderingTask {
+    pub async fn run(pool: SqlitePool, q_depth: usize, load_delivered: bool) -> DeliverableTask {
+        let (tx, rx) = tokio::sync::mpsc::channel::<DeliverableEvent>(q_depth);
+        let (tx_new, rx_new) = tokio::sync::mpsc::channel::<DeliveredEvent>(q_depth);
+
+        let handle =
+            tokio::spawn(async move { Self::run_loop(pool, load_delivered, rx, rx_new).await });
+
+        DeliverableTask {
+            _handle: handle,
+            tx,
+            tx_new,
+        }
+    }
+
+    async fn run_loop(
+        pool: SqlitePool,
+        load_undelivered: bool,
+        mut rx: tokio::sync::mpsc::Receiver<DeliverableEvent>,
+        mut rx_new: tokio::sync::mpsc::Receiver<DeliveredEvent>,
+    ) {
+        // before starting, make sure we've updated any events in the database we missed
+        let mut state = OrderingState::new();
+        if load_undelivered
+            && state
+                .process_all_undelivered_events(&pool, MAX_ITERATIONS)
+                .await
+                .map_err(Self::log_error)
+                .is_err()
+        {
+            return;
+        }
+
+        loop {
+            let mut need_prev_buf = Vec::with_capacity(100);
+            let mut newly_added_buf = Vec::with_capacity(100);
+
+            tokio::select! {
+                incoming = rx.recv_many(&mut need_prev_buf, 100) => {
+                    if incoming > 0 {
+                        state.add_incoming_batch(need_prev_buf);
+                    }
+                }
+                new = rx_new.recv_many(&mut newly_added_buf, 100) => {
+                    if new > 0 {
+                        let mut new = false;
+                        for item in newly_added_buf {
+                            if let Some(waiting) = state.pending_by_stream.get_mut(&item.init_cid) {
+                                if let Some(good_to_go) = waiting.remove_by_prev_cid(&item.cid) {
+                                    state.ready_events.push_back(good_to_go);
+                                    tracing::trace!(%good_to_go, "Found event unblocked by incoming delivered list.");
+                                    new = true;
+                                }
+                            }
+                        }
+                        if new && state.persist_ready_events(&pool).await.map_err(Self::log_error).is_err() {
+                            return;
+                        }
+                    }
+                }
+                else => {
+                    info!("Server dropped the ordering task. Processing once more before exiting...");
+                    let _ = state
+                        .process_events(&pool)
+                        .await
+                        .map_err(Self::log_error);
+                    return;
+                }
+            };
+
+            if state
```

Comment:
Remove this or do it very infrequently. We should process anything blocked when we hear about it through the `rx_new` channel. 

As thing get worse (i.e. we have a ton of streams we can't deliver), this will loop over them all doing a lot of pointless work we can't make progress on. It may catch things that were missed somehow, but there are better ways to deal with this.